File: julia_solutions/day03/day03.jl
module Day03

export run

# Function for the problem
function run(input::String)
    part_1 = part1(input)
    part_2 = part2(input)
    return [part_1, part_2]
end

# Function for part1 of the problem
function part1(input::String)
    NOT_SYMBOLS = ['.', '1', '2', '3', '4', '5', '6', '7', '8', '9', '0']
    DIRECTIONS = [[1,0], [-1, 0], [0,1], [0,-1], [1,1], [1,-1], [-1,1], [-1,-1]]
    lines = readlines(input) #Reading input data
    count = 0
    for (i, line) in enumerate(lines)
        ranges =  findall(r"(\d+)", line)
        for range in ranges
            surrounded = false
            for j in range
                surrounded = [!(lines[dir[1]][dir[2]] in NOT_SYMBOLS) for dir in DIRECTIONS .+ [[i,j]] if 1 <= dir[1] < length(lines) && 1 <= dir[2] < length(line)] |> any
                if surrounded == true
                    count += parse(Int, prod(line[range]))
                    break
                end
            end
        end
    end
    return count
end

# Function for part2 of the problem
function part2(input::String)
    DIRECTIONS = [[1,0], [-1, 0], [0,1], [0,-1], [1,1], [1,-1], [-1,1], [-1,-1]]
    lines = readlines(input) #Reading input data
    dic = Dict()
    for (i, line) in enumerate(lines)  
        ranges =  findall(r"(\d+)", line)
        for range in ranges
            surrounded = false 
            for j in range
                for pos in DIRECTIONS .+ [[i,j]]
                    if 1 <= pos[1] < length(lines) && 1 <= pos[2] < length(line) && lines[pos[1]][pos[2]]=='*'
                        try
                            append!(dic[(pos[1],pos[2])],parse(Int, prod(line[range])))
                            surrounded = true
                        catch
                            dic[(pos[1],pos[2])]=[parse(Int, prod(line[range]))]
                            surrounded = true
                        end
                        break
                    end
                end
                surrounded && break 
            end
        end
    end
    return [value[1]*value[2] for (_, value) in dic if length(value)==2] |> sum
end

end

File: golang_solutions/day03/day03.go
package day03

import (
	"log"
	"os"
	"regexp"
	"slices"
	"strconv"
	"strings"
)

func checkSurrounded(nline int, match []int, lines []string) bool {
	notSymbols := []string{".", "1", "2", "3", "4", "5", "6", "7", "8", "9", "0"}
	surrounded := false
	for ncol := match[0]; ncol < match[1]; ncol++ {
		switch {
		case nline > 0 && !slices.Contains(notSymbols, string(lines[nline-1][ncol])):
			surrounded = true
		case nline < len(lines)-1 && !slices.Contains(notSymbols, string(lines[nline+1][ncol])):
			surrounded = true
		case ncol > 0 && !slices.Contains(notSymbols, string(lines[nline][ncol-1])):
			surrounded = true
		case ncol < len(lines[nline])-1 && !slices.Contains(notSymbols, string(lines[nline][ncol+1])):
			surrounded = true
		case nline > 0 && ncol > 0 && !slices.Contains(notSymbols, string(lines[nline-1][ncol-1])):
			surrounded = true
		case nline > 0 && ncol < len(lines[nline])-1 && !slices.Contains(notSymbols, string(lines[nline-1][ncol+1])):
			surrounded = true
		case nline < len(lines)-1 && ncol > 0 && !slices.Contains(notSymbols, string(lines[nline+1][ncol-1])):
			surrounded = true
		case nline < len(lines)-1 && ncol < len(lines[nline])-1 && !slices.Contains(notSymbols, string(lines[nline+1][ncol+1])):
			surrounded = true
		}

	}
	return surrounded
}

func checkGear(nline int, match []int, lines []string) (bool, [][2]int) {
	notSymbols := []string{".", "1", "2", "3", "4", "5", "6", "7", "8", "9", "0"}
	gear := false
	pos := [][2]int{}
	for ncol := match[0]; ncol < match[1]; ncol++ {
		switch {
		case nline > 0 && string(lines[nline-1][ncol]) == "*":
			gear = true
			pos = append(pos, [2]int{nline - 1, ncol})
		case nline < len(lines)-1 && string(lines[nline+1][ncol]) == "*":
			gear = true
			pos = append(pos, [2]int{nline + 1, ncol})
		case ncol > 0 && string(lines[nline][ncol-1]) == "*":
			gear = true
			pos = append(pos, [2]int{nline, ncol - 1})
		case ncol < len(lines[nline])-1 && string(lines[nline][ncol+1]) == "*":
			gear = true
			pos = append(pos, [2]int{nline, ncol + 1})
		case nline > 0 && ncol > 0 && !slices.Contains(notSymbols, string(lines[nline-1][ncol-1])):
			gear = true
			pos = append(pos, [2]int{nline - 1, ncol - 1})
		case nline > 0 && ncol < len(lines[nline])-1 && !slices.Contains(notSymbols, string(lines[nline-1][ncol+1])):
			gear = true
			pos = append(pos, [2]int{nline - 1, ncol + 1})
		case nline < len(lines)-1 && ncol > 0 && !slices.Contains(notSymbols, string(lines[nline+1][ncol-1])):
			gear = true
			pos = append(pos, [2]int{nline + 1, ncol - 1})
		case nline < len(lines)-1 && ncol < len(lines[nline])-1 && !slices.Contains(notSymbols, string(lines[nline+1][ncol+1])):
			gear = true
			pos = append(pos, [2]int{nline + 1, ncol + 1})
		}
		if gear == true {
			break
		}
	}
	return gear, pos
}

func Part1(filename string) int {

	input, err := os.ReadFile(filename)

	if err != nil {
		log.Fatal("File not found")
	}

	lines := strings.Split(string(input), "\n")

	sum := 0
	for i, line := range lines {
		matches := regexp.MustCompile(`(\d+)`).FindAllStringIndex(line, -1)
		for _, match := range matches {
			if checkSurrounded(i, match, lines) {
				number, _ := strconv.Atoi(line[match[0]:match[1]])
				sum = sum + number
			}
		}
	}

	return sum

}

func Part2(filename string) int {

	input, err := os.ReadFile(filename)

	if err != nil {
		log.Fatal("File not found")
	}

	lines := strings.Split(string(input), "\n")

	gearsDict := make(map[[2]int][]int)
	for i, line := range lines {
		matches := regexp.MustCompile(`(\d+)`).FindAllStringIndex(line, -1)
		for _, match := range matches {
			check, gearPositions := checkGear(i, match, lines)
			if check {
				number, _ := strconv.Atoi(line[match[0]:match[1]])
				for _, position := range gearPositions {
					gearsDict[position] = append(gearsDict[position], number)
				}
			}
		}
	}

	sum := 0
	for _, gear := range gearsDict {
		if len(gear) == 2 {
			sum = sum + gear[0]*gear[1]
		}
	}

	return sum

}
